### PR TITLE
Support Organizations & Locations metadata

### PR DESCRIPTION
**Pull request by sean797**
This adds support for Organizations & Locations metadata. 
You can test with metadata:
```
locations:
- loc a
- loc b
organizations:
- org a
- org b
```

### CODE REVIEW

File: app/models/concerns/foreman_templates/ptable_import.rb [hunk 1]
```diff
@@ -38,11 +47,22 @@ def import!(name, text, metadata)
     end
 
     # TODO: DRY this, it's copied from ProvisioningTemplateImport
-    def map_oses(metadata)
```
Comment (GregSutcliffe) [inline app/models/concerns/foreman_templates/ptable_import.rb]: I like the refactoring on this method - but we should probably handle this TODO as well. Could you move this method to the TemplateImporter baseclass concern? It should then be available to both of these subclasses.

File: app/models/concerns/foreman_templates/provisioning_template_import.rb [hunk 2]
```diff
@@ -36,11 +40,15 @@ def import!(name, text, metadata)
         status  = template.update_attributes(data)
         result  = "  #{c_or_u} Template #{id_string}:#{name}"
         result += "\n    Operatingsystem Associations:\n    - #{oses.map(&:fullname).join("\n    - ")}" unless oses.empty?
-      elsif data[:operatingsystem_ids]
+        result += "\n    Organizations Associations:\n    - #{organizations.map(&:name).join("\n    - ")}" unless organizations.empty?
+        result += "\n    Location Associations:\n    - #{locations.map(&:name).join("\n    - ")}" unless locations.empty?
+      elsif data[:operatingsystem_ids] || data[:location_ids] || data[:organization_ids]
```
Comment (GregSutcliffe) [inline app/models/concerns/foreman_templates/provisioning_template_import.rb]: This is starting to feel a bit "spammy" - but I don't have any good solutions right now. `result` is only shown in verbose mode anyway, so it's probably ok for now - but suggestions on refactoring the output from the rake task are welcome ;)

File: app/models/concerns/foreman_templates/ptable_import.rb (hunk 1 above)
Comment (GregSutcliffe) [inline app/models/concerns/foreman_templates/ptable_import.rb]: This would simplify the tests too ;)

File: test/unit/ptable_test.rb [hunk 3]
```diff
@@ -101,20 +101,48 @@ def setup
...
+
+      test 'map_metadata defaults to an emtpy array' do
         metadata = {}
-        assert_equal [], Ptable.map_oses(metadata)
+        [ 'oses', 'locations', 'organizations' ].each do |param|
+          assert_equal [], Ptable.map_metadata(metadata, param)
+        end
       end
     end
   end
```
Comment (GregSutcliffe) [inline test/unit/ptable_test.rb]: I don't see any additions to the tests to show a template actually gets a taxonomy assigned?

File: test/unit/ptable_test.rb [hunk 4]
```diff
@@ -49,6 +52,7 @@ class PtableTest < ActiveSupport::TestCase
 
         pt = Ptable.find_by_name(name)
         assert_equal @os_deb.family, pt.os_family
+        assert_equal @new_org.id, pt.organization_ids
```
Comment (GregSutcliffe) [inline test/unit/ptable_test.rb]: This test is failing for me, `pt.organization_ids` is `[]` - moreover, you probably want `[@new_org.id]` as the first argument, else you're comparing Integer to Array.

File: test/unit/ptable_test.rb [hunk 5]
```diff
@@ -4,16 +4,19 @@ module ForemanTemplates
...
     setup do
       # create a basic template with no OS assigned
-      @os_rh  = FactoryGirl.create(:operatingsystem, :family => 'Redhat')
-      @os_deb = FactoryGirl.create(:operatingsystem, :family => 'Debian')
-      @pt     = FactoryGirl.create(:ptable, :os_family => 'Redhat')
+      @os_rh   = FactoryGirl.create(:operatingsystem, :family => 'Redhat')
+      @os_deb  = FactoryGirl.create(:operatingsystem, :family => 'Debian')
+      @pt      = FactoryGirl.create(:ptable, :os_family => 'Redhat')
+      @new_org = FactoryGirl.create(:organization, :name => 'NewOrg')
+      @old_org = FactoryGirl.create(:organization, :name => 'OldOrg')
```
Comment (GregSutcliffe) [inline test/unit/ptable_test.rb]: You need this in provisioning_template_test too - in fact it's more important there :P

File: test/unit/ptable_test.rb (hunk 4 above)
Comment (GregSutcliffe) [inline test/unit/ptable_test.rb]: Oh, and this should probably be in a new test - or at least reword the title to mention orgs.

File: test/unit/provisioning_template_test.rb [hunk 6]
```diff
@@ -141,21 +141,49 @@ def setup
...
+      test 'returns Locations that are in the db' do
+        metadata = { 'locations' => ['Location 1', 'Location One'] }
+        assert_equal [Location.find_by_name('Location 1')],
+                     ProvisioningTemplate.map_metadata(metadata, 'locations')
       end
 
       test 'returns an empty array for no matched OSes' do
         metadata = { 'oses' => ['Fedora 19'] }
-        assert_equal [], ProvisioningTemplate.map_oses(metadata)
+        assert_equal [], ProvisioningTemplate.map_metadata(metadata, 'oses')
```
Comment (GregSutcliffe) [inline test/unit/provisioning_template_test.rb]: can we DRY these tests in a new TemplateTest class? Rather than copying these changes to both files...

**Comment by sean797** [comment]
Okay, I've made those changes and test are passing locally, should be good to merge - hopefully :+1:

**Comment by sean797** [comment]
ping @GregSutcliffe sorry to ask but any chance you can review this again or merge? We want to use this in production but ideally wanted to wait until it was merged to do so.

**Comment by GregSutcliffe** [comment]
@sean797 my apologies, this one slipped through. I'm on conference today but I'll try to take a look later this week.

File: app/models/concerns/foreman_templates/provisioning_template_import.rb (hunk 2 above)
Comment (ares) [inline app/models/concerns/foreman_templates/provisioning_template_import.rb]: probably better to use standard logger with different log levels, built-in ruby logger should be enough but I guess it's out of scope for this PR

Comment (GregSutcliffe) [inline app/models/concerns/foreman_templates/provisioning_template_import.rb]: Yeah, it needs to either go to logs for calling within the app, or to stdout when called from Rake. Needs a rethink on output handling in general, really. For now, it'll be fine.

**Comment by GregSutcliffe** [comment]
Merged as 1b55133 with a few minor whitespace/syntax tidyups - thanks @sean797!

**Pull request closed without merging**